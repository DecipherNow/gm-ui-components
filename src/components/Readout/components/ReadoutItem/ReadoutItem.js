import React from "react";
import { PropTypes } from "prop-types";

import * as Glyphs from "../../../Glyphs";
import ItemDisplay from "./components/ItemDisplay";
import ReadoutItemData from "./components/ReadoutItemData";
import ReadoutItemGraph from "./components/ReadoutItemGraph";
import ReadoutItemIcon from "./components/ReadoutItemIcon";
import ReadoutItemTitle from "./components/ReadoutItemTitle";
import ReadoutItemValue from "./components/ReadoutItemValue";
import ReadoutItemDetail from "./components/ReadoutItemDetail";

//  TO-DO:  icon name, shape, color and other icon specifics are to be passed to getIcon component that renders svg element
// when we have "future", "bolt", "warning", "server", "link" svg elements...

export default function ReadoutItem({
  children,
  detail,
  graphData,
  icon,
  iconName,
  iconBackgroundStyle,
  iconBorderStyle,
  iconBorderWidth,
  iconShape,
  iconColor,
  title,
  value
}) {
  let GlyphComponent;
  if (icon) {
    GlyphComponent = Glyphs[icon];
  }
  return (
    <ItemDisplay>
      <ReadoutItemData>
        <ReadoutItemTitle>
          {icon && (
            <ReadoutItemIcon
              iconBackgroundStyle={iconBackgroundStyle}
              iconBorderStyle={iconBorderStyle}
              iconBorderWidth={"1"}
            >
              <GlyphComponent />
            </ReadoutItemIcon>
          )}
          {title || "—"}
        </ReadoutItemTitle>
        <ReadoutItemValue>{value || "—"}</ReadoutItemValue>
        {detail && <ReadoutItemDetail>{detail || "—"}</ReadoutItemDetail>}
        {children}
      </ReadoutItemData>
    </ItemDisplay>
  );
}

const readoutItemPropTypes = {
  children: PropTypes.element,
  detail: PropTypes.string,
  graphData: PropTypes.arrayOf(PropTypes.number),
  icon: PropTypes.string,
  iconBackgroundStyle: PropTypes.string,
  iconBorderStyle: PropTypes.string,
  iconBorderWidth: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  iconColor: PropTypes.string,
  iconName: PropTypes.string,
  iconShape: PropTypes.string,
  title: PropTypes.string,
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
    PropTypes.object
  ])
};

export const ReadoutItemShape = PropTypes.shape(readoutItemPropTypes);
ReadoutItem.propTypes = readoutItemPropTypes;
