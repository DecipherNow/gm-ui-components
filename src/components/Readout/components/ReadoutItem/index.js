import ReadoutItem, { ReadoutItemShape } from "./ReadoutItem";
export default ReadoutItem;
export { ReadoutItemShape };
