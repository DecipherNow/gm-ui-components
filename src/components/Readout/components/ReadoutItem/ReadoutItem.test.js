import React from "react";

import ReadoutItem from "./ReadoutItem";
import { IconAlertTriangle } from "../../../";

describe("ReadoutItem", () => {
  it("matches snapshot", () => {
    let aReadoutItem = shallow(
      <ReadoutItem title={"Readout"} value="100" detail="Readout Detail" />
    );
    expect(aReadoutItem).toMatchSnapshot();
  });
  it("renders an icon if provided", () => {
    let aReadoutItem = shallow(
      <ReadoutItem
        title={"Readout"}
        value="100"
        detail="Readout Detail"
        icon={() => <IconAlertTriangle />}
      />
    );
    expect(aReadoutItem.find(IconAlertTriangle)).toHaveLength(1);
    aReadoutItem = shallow(<ReadoutItem />);
    expect(aReadoutItem.find(IconAlertTriangle)).toHaveLength(0);
    expect(aReadoutItem.html().includes("-")).toBe(true);
  });
  it("renders a title if provided, and a - if not", () => {
    let aReadoutItem = shallow(<ReadoutItem title="A Super Cool Readout" />);
    expect(
      aReadoutItem
        .find("ReadoutItemTitle")
        .html()
        .includes("A Super Cool Readout")
    ).toBe(true);
    aReadoutItem = shallow(<ReadoutItem />);
    expect(
      aReadoutItem
        .find("ReadoutItemTitle")
        .html()
        .includes("A Super Cool Readout")
    ).toBe(false);
  });
  it("renders a value if provided, and a - if not", () => {
    let aReadoutItem = shallow(<ReadoutItem value="100" />);
    expect(aReadoutItem.find("ReadoutItemValue").html().includes("100")).toBe(
      true
    );
    aReadoutItem = shallow(<ReadoutItem />);
    expect(aReadoutItem.find("ReadoutItemValue").html().includes("100")).toBe(
      false
    );
  });
  it("renders a detail if provided, and nothing if not", () => {
    let aReadoutItem = shallow(
      <ReadoutItem detail="Is Justin Ely a better developer than Sean McBride?" />
    );
    expect(
      aReadoutItem.find("ReadoutItemDetail").html().includes("Justin Ely")
    ).toBe(true);
    aReadoutItem = shallow(<ReadoutItem />);
    expect(aReadoutItem.find("ReadoutItemDetail")).toHaveLength(0);
  });
});
