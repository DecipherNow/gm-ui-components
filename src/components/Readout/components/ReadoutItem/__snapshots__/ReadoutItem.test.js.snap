// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReadoutItem matches snapshot 1`] = `
<ItemDisplay>
  <ReadoutItemData>
    <ReadoutItemTitle>
      Readout
    </ReadoutItemTitle>
    <ReadoutItemValue>
      100
    </ReadoutItemValue>
    <ReadoutItemDetail>
      Readout Detail
    </ReadoutItemDetail>
  </ReadoutItemData>
</ItemDisplay>
`;
